### PR TITLE
Refs #32365 -- Allowed use of non-pytz timezone implementations.

### PR DESCRIPTION
**Pull request by pganssle**
The current plan for switching Django over to use the `zoneinfo` module is a hard switch in Django 4.0, but in the meantime it would be prudent to make it at least _possible_ to use `zoneinfo` with Django. I am not sure if this will be a perfectly smooth experience for the end user, but I've added `zoneinfo`-based tests alongside (nearly) all `pytz`-based tests that I could find, then fixed any resulting errors.

I know the deadline is very tight on this so I did not try to get too fancy with any formatting or the structures here so that we could get the review process started as soon as possible.

If this approach is acceptable, I think I still also need to add a few tests to:

- [x] `tests/admin_widgets`: unnecessary — `pytz` is not used for anything consequential
- [x]  `tests/admin_views`: done
- [x] `tests/db_functions/test_extract_trunc` (truncating with ambiguous and imaginary times — this may need a logic change as well).

CC: @carltongibson @adamchainz

### CODE REVIEW

File: tests/timezones/tests.py [hunk 1]
```diff
@@ -7,6 +7,14 @@
 
 import pytz
 
+try:
```
Comment (pganssle) [inline tests/timezones/tests.py:10]: Note: This idiom gets repeated in a bunch of places, not sure if there's a "test support" module or something you want me to put it in. I frequently bury this sort of thing in a `_compat` module somewhere.

File: django/utils/timezone.py [hunk 2]
```diff
@@ -191,7 +191,7 @@ def now():
     """
     if settings.USE_TZ:
         # timeit shows that datetime.now(tz=utc) is 24% slower
```
Comment (pganssle) [inline django/utils/timezone.py]: This is technically an unrelated change, but per [this blog post](https://blog.ganssle.io/articles/2019/11/utcnow.html), I don't think it's a good idea in general to use`.utcnow`. From my tests, `%timeit datetime.now(tz=pytz.UTC)` is approximately as fast as `datetime.utcnow().replace(tzinfo=utc)` now, and when `utc` starts pointing to `datetime.timezone.utc` instead of `pytz.UTC`, `datetime.now(utc)` will be *dramatically* faster, so may as well make that upgrade seamless while we're at it.

If you want me to move this to a separate PR please let me know.

Comment (carltongibson) [reply django/utils/timezone.py]: Perhaps just a separate commit is fine.

**Comment by carltongibson** [comment]
Hey @pganssle — thanks for this! 

I've [posted to the list about whether we can squeeze this in, if we can't get full review before Thursday](https://groups.google.com/g/django-developers/c/cXIyCcjMOds/m/L1H2hlHpCgAJ). 

Are the TODOs in your first comment still up to date?

File: tests/timezones/tests.py (hunk 1 above)
Comment (charettes) [reply tests/timezones/tests.py:10]: I guess we could introduce `django.utils.zoneinfo` like we did with `simplejson` back in the time.

We could start deprecating the module when we drop support for Python 3.8.

**Comment by pganssle** [comment]
@carltongibson I have updated the TODOs accordingly. I think that there are two remaining questions here:

1. Currently, [`make_aware`](https://github.com/django/django/blob/8707e3f2dd07ab00ff33849615017c7a13bb64c6/django/utils/timezone.py#L227-L240) will ignore `is_dst` for non-`pytz` zones, because [the semantics of `fold` and `is_dst` is not a perfect mapping](https://pytz-deprecation-shim.readthedocs.io/en/latest/migration.html#semantic-differences-between-is-dst-and-fold). Since `fold` is set on the `value` parameter and has no default, there's no way to detect if someone has deliberately set `fold` *and* `is_dst`, and it's not clear which should take precedence.

I could reasonably migrate the [logic from `pytz-deprecation-shim`](https://github.com/pganssle/pytz-deprecation-shim/blob/47bd4bdd9346cafa6c6d66817082ccce099890ad/src/pytz_deprecation_shim/_impl.py#L220-L263) here, which is that `is_dst` is ignored when not specified (`value`'s fold is used instead), and if explicitly specified, normal `pytz` logic applies (we'd need to swap out the `AmbiguousTimeError` and `NonExistentTimeError` for another exception type in the `None` case, obviously). In the end I think `is_dst` should be deprecated or removed or something in whatever version built-in `zoneinfo` support rolls out in, since the semantics aren't quite right.

2. A related concern is what to do about [this date truncation logic](https://github.com/django/django/blob/8707e3f2dd07ab00ff33849615017c7a13bb64c6/tests/db_functions/datetime/test_extract_trunc.py#L1192-L1214). This is similar to the other issue in that it comes down to how `is_dst` should work. The difference is that with `make_aware`, you are passing a `datetime` directly, and can thus set the fold. In this case the datetime is created for you behind the scenes, and you only have the `is_dst` lever. We can leave this as-is and have people "fix it in post" if they opt-in, or we can add the `is_dst` resolution logic directly into the `TruncBase.convert_value` code, or we can add the `is_dst` support to `make_aware` and the truncation logic will be inherited automatically.

That stuff is the only possible backwards-compatibility concern in this PR, since to the extent that it's possible to use `make_aware` and the truncation logic with non-`pytz` zones _today_, `is_dst` is already ignored. Making it sometimes raise exceptions could be a backwards incompatible change (though if we're taking something that was broken and unusable and making it usable maybe that's not a really incompatible change). One option to get around this would be to disable the `is_dst=None` behavior for non-`pytz` zones and add a new singleton like `django.utils.timezone.ERROR_UNLESS_UNAMBIGUOUS` (or accept the string `'error'`) that triggers the `is_dst=None` behavior even in non-`pytz` zones.

In any case, I would say that even without any changes to the `is_dst` logic, this PR is a win for DJango users, and we shouldn't let the perfect be the enemy of the good. I think any wonky DST logic that would come about from the lack of `is_dst` support for non-`pytz` zones could be fixed reasonably easily by end users anyway.

File: tests/timezones/tests.py (hunk 1 above)
Comment (pganssle) [reply tests/timezones/tests.py:10]: I wouldn't want anything user-facing here, since this is only ever used in the tests. If there's a common test utility module somewhere in the test suite where I can add functions like, "Get this IANA key in all supported time zone providers" and "skip this test if zoneinfo isn't available", it would make the PR a bit cleaner, but probably a lot of this stuff will get refactored for 4.0 anyway, so I don't think it's critical.

File: tests/utils_tests/test_timezone.py [hunk 3]
```diff
@@ -1,14 +1,38 @@
...
+
+HAS_ZONEINFO = zoneinfo is not None
+PARIS_IMPLS = (PARIS,)
+
+if not HAS_ZONEINFO:
+    PARIS_ZI = None
+    PARIS_IMPLS = (PARIS,)
+
+    def needs_zoneinfo(f):
+        return f
```
Comment (adamchainz) [inline tests/utils_tests/test_timezone.py:31]: isn't this back to front? The skip definition should be here, for if the user doesn't have zoneinfo?

File: tests/utils_tests/test_timezone.py [hunk 4]
```diff
@@ -1,14 +1,38 @@
...
 from django.utils import timezone
 
-CET = pytz.timezone("Europe/Paris")
+PARIS = pytz.timezone('Europe/Paris')
 EAT = timezone.get_fixed_timezone(180)      # Africa/Nairobi
 ICT = timezone.get_fixed_timezone(420)      # Asia/Bangkok
+UTC = datetime.timezone.utc
+
+HAS_ZONEINFO = zoneinfo is not None
+PARIS_IMPLS = (PARIS,)
```
Comment (adamchainz) [inline tests/utils_tests/test_timezone.py:24]: no need to define here if it's defined by the below `if`

File: django/utils/timezone.py [hunk 5]
```diff
@@ -249,3 +248,29 @@ def make_naive(value, timezone=None):
     if is_naive(value):
         raise ValueError("make_naive() cannot be applied to a naive datetime")
     return value.astimezone(timezone).replace(tzinfo=None)
+
+
+_PYTZ_BASE_CLASSES = None
```
Comment (adamchainz) [inline django/utils/timezone.py:253]: why not define this tuple at import time? `import pytz` seems to already have made the classes accessible.

File: django/utils/timezone.py [hunk 6]
```diff
@@ -249,3 +248,29 @@ def make_naive(value, timezone=None):
...
+    global _PYTZ_BASE_CLASSES
+    if _PYTZ_BASE_CLASSES is None:
+        _PYTZ_BASE_CLASSES = (pytz.tzinfo.BaseTzInfo, pytz._FixedOffset)
+        if not isinstance(pytz.UTC, pytz._FixedOffset):
+            _PYTZ_BASE_CLASSES = _PYTZ_BASE_CLASSES + (type(pytz.UTC),)
+
+    return isinstance(tz, _PYTZ_BASE_CLASSES)
+
+
+def _datetime_unambiguous(dt, tz):
```
Comment (adamchainz) [inline django/utils/timezone.py:267]: This function is mostly used with a `not`. To avoid the double negative, could we swap it and rename it to `_datetime_ambiguous`, so we'd have e.g. `if _datetime_ambiguous(...)` ?

**Review by adamchainz — COMMENTED** [review]
There are some pytz references in `docs/i18n/timezones.txt` that should likely be updated to be about "your timezone provider"

File: django/utils/dateformat.py [hunk 7]
```diff
@@ -160,16 +161,11 @@ def T(self):
         if not self.timezone:
             return ""
 
-        name = None
-        try:
+        if _datetime_unambiguous(self.data, tz=self.timezone):
```
Comment (pauloxnet) [inline django/utils/dateformat.py:164]: Is `tz` necessary here ? elsewhere `self.timezone` is passed without it.

**Review by pauloxnet — APPROVED** [review]

**Comment by pauloxnet** [comment]
It seems ok to me. I added only a small comment.

File: django/utils/timezone.py (hunk 6 above)
Comment (pganssle) [reply django/utils/timezone.py:263]: We could do this, but there's an unfortunate fuzziness to the terminology here, as "ambiguous" is something of a term of art meaning "this datetime occurs twice in the zone", but this function detects any instance when there are two different possible offsets for a given naïve datetime  — which is to say that "unambiguous" in this context is `not (datetime_ambiguous(dt) or datetime_imaginary(dt))` rather than just `not datetime_ambiguous(dt)`.

I am not quite sure how to resolve this issue with the nomenclature, but I welcome suggestions. 😅 I don't think there's a specific word for "not ambiguous and not imaginary". [PEP 495](https://www.python.org/dev/peps/pep-0495/#terminology) does not seem to suggest a word, but I haven't re-read it carefully lately.

File: tests/utils_tests/test_timezone.py (hunk 3 above)
Comment (pganssle) [reply tests/utils_tests/test_timezone.py]: Oops, good catch!

File: django/utils/dateformat.py (hunk 7 above)
Comment (pganssle) [reply django/utils/dateformat.py:164]: I guess that's a style question. Happy to consistently pass by keyword or by position. For now, I'll remove the `tz=` and if y'all want me to switch to passing by keyword everywhere I'll do that.

File: django/utils/timezone.py (hunk 5 above)
Comment (pganssle) [reply django/utils/timezone.py]: This was originally part of `pytz-deprecation-shim` and had all kinds of guards to avoid any unnecessary access to `pytz`. When I first migrated it to Django, it was in another file that didn't directly import `pytz`, but you're right that there's no point trying to be lazy in `django.utils.timezone`, so I've gone ahead and moved it to populating at import time.

File: tests/timezones/tests.py (hunk 1 above)
Comment (charettes) [reply tests/timezones/tests.py:10]: right then I guess this import sequence is not too bad, it's already something we do in tests in a few other places when dealing with optional libraries.

**Comment by pganssle** [comment]
@adamchainz I've made some minor updates to the i18n documentation, but a lot of the references to `pytz` are very specific to `pytz`, and might require a more significant overhaul to show both ways of doing things.

I'd also say that [this advice](https://docs.djangoproject.com/en/3.1/topics/i18n/timezones/#selecting-the-current-time-zone), while not so `pytz`-specific, is a bit of a problematic recommendation, as the `tz` project is pretty clear that the database keys that `pytz.common_timezones` and `zoneinfo.available_timezones()` expose are not really intended to be user-facing, but "how to create a timezone-picker" is a much more difficult and complicated topic to address.

I think for now it's probably best to just acknowledge that non-`pytz` zones are supported (and intended to be supported), and decide later on whether it's worth overhauling the docs to show both `pytz` and `zoneinfo` support side-by-side.

File: django/utils/dateformat.py (hunk 7 above)
Comment (pauloxnet) [reply django/utils/dateformat.py]: Yes, it was only minor style comment. Thanks for the update.

**Review by aaugustin — APPROVED** [review]
Sorry I don't have time to perform a full review. At a high level this looks OK to me. Thanks Paul & everyone else involved!

File: django/utils/timezone.py (hunk 6 above)
Comment (aaugustin) [reply django/utils/timezone.py:263]: I'd be OK with calling the function `_datetime_ambiguous_or_imaginary`.

It's verbose, but anything shorter will be confusing, given what Paul said.

File: django/utils/timezone.py (hunk 2 above)
Comment (aaugustin) [reply django/utils/timezone.py]: My micro-optimizations from 2011 are so cute :-)

https://github.com/django/django/commit/9b1cb755a28f020e27d4268c214b25315d4de42e#diff-698cf55de9163a8fc8381fc3f0320d8336203c45e03306f497c67495bffce40bR224

Then at some point I was the Python 3 guy but I wasn't aware of that subtlety with the datetime module!

Thanks for cleaning that up.

File: django/utils/timezone.py (hunk 6 above)
Comment (pganssle) [reply django/utils/timezone.py]: OK, I've made that change.

File: tests/timezones/tests.py [hunk 8]
```diff
@@ -51,6 +61,13 @@
 EAT = timezone.get_fixed_timezone(180)      # Africa/Nairobi
 ICT = timezone.get_fixed_timezone(420)      # Asia/Bangkok
 
+PARIS_IMPLS = [
```
Comment (carltongibson) [inline tests/timezones/tests.py:64]: Hey @pganssle — I'm being slow (as ever 🙂) what's `IMPL` stand for here? (Ta!)

Comment (pganssle) [reply tests/timezones/tests.py:64]: Yeah, implementation. I could also change this to be something more like I did in the other tests, a `def get_timezones(key)` function that returns a list of the specific implementations rather than using a global like this.

Comment (pganssle) [reply tests/timezones/tests.py]: I went ahead and made that change in a fixup commit so you can see what it looks like. Let me know what you want to do and I'll either squash it down or drop the new commit.

File: docs/topics/i18n/timezones.txt [hunk 9]
```diff
@@ -26,8 +26,11 @@ to this problem is to use UTC in the code and use local time only when
 interacting with end users.
 
 Time zone support is disabled by default. To enable it, set :setting:`USE_TZ =
-True <USE_TZ>` in your settings file. Time zone support uses pytz_, which is
-installed when you install Django.
+True <USE_TZ>` in your settings file. By default, time zone support uses pytz_,
+which is installed when you install Django; Django also supports the use of
+other time zone implementations like zoneinfo_ by passing
```
Comment (wkschwartz) [inline docs/topics/i18n/timezones.txt:31]: I think you can do without hard coding the link by using
```suggestion
other time zone implementations like :mod:`zoneinfo` by passing
```
(If you do this, just don't forget to delete the `.. _zoneinfo:` line at the end)

Comment (pganssle) [reply docs/topics/i18n/timezones.txt]: Ah yeah, originally I wasn't sure if this sometimes renders with references to the Python < 3.8 documentation (in which case it would be a dead link), but considering I use a `:func:` reference elsewhere I think it's a safe bet to go with this.

Comment (wkschwartz) [reply docs/topics/i18n/timezones.txt]: Renders with current Python, i.e., >= 3.9:

https://github.com/django/django/blob/e0a46367df8b17905f1c78f5c86f88d21c0f2b4d/docs/conf.py#L148-L149

File: django/forms/utils.py [hunk 10]
```diff
@@ -158,10 +158,14 @@ def from_current_timezone(value):
...
     """
+    return_value = value
     if settings.USE_TZ and value is not None and timezone.is_naive(value):
         current_timezone = timezone.get_current_timezone()
         try:
-            return timezone.make_aware(value, current_timezone)
+            return_value = timezone.make_aware(value, current_timezone)
+            if not (timezone._is_pytz_zone(current_timezone) or
+                    not timezone._datetime_ambiguous_or_imaginary(value, current_timezone)):
+                raise ValueError('Ambiguous or non-existent time.')
```
Comment (felixxm) [inline django/forms/utils.py:170]: Do we have test for this error?

File: django/forms/utils.py [hunk 11]
```diff
@@ -158,10 +158,16 @@ def from_current_timezone(value):
...
     """
+    return_value = value
     if settings.USE_TZ and value is not None and timezone.is_naive(value):
         current_timezone = timezone.get_current_timezone()
         try:
-            return timezone.make_aware(value, current_timezone)
+            return_value = timezone.make_aware(value, current_timezone)
+            if (
+                not timezone._is_pytz_zone(current_timezone) and
+                timezone._datetime_ambiguous_or_imaginary(value, current_timezone)
```
Comment (felixxm) [inline django/forms/utils.py:168]: `value` or `return_value`? or maybe we should swap these lines:
```python
if (
    not timezone._is_pytz_zone(current_timezone) and
    timezone._datetime_ambiguous_or_imaginary(value, current_timezone)
):
    raise ValueError('Ambiguous or non-existent time.')
```

File: django/utils/dateformat.py [hunk 12]
```diff
@@ -160,16 +161,11 @@ def T(self):
...
             name = self.timezone.tzname(self.data)
-        except Exception:
-            # pytz raises AmbiguousTimeError during the autumn DST change.
-            # This happens mainly when __init__ receives a naive datetime
-            # and sets self.timezone = get_default_timezone().
-            pass
-        if name is None:
+        else:
             name = self.format('O')
+
```
Comment (felixxm) [inline django/utils/dateformat.py:168]: Chop blank line.

File: django/utils/dateformat.py [hunk 13]
```diff
@@ -184,16 +180,11 @@ def Z(self):
 
         If timezone information is not available, return an empty string.
         """
-        if not self.timezone:
+        if (not self.timezone or
+                _datetime_ambiguous_or_imaginary(self.data, self.timezone)):
```
Comment (felixxm) [inline django/utils/dateformat.py:184]:
```suggestion
        if (
            not self.timezone or
            _datetime_ambiguous_or_imaginary(self.data, self.timezone)
        ):
```

File: django/utils/dateformat.py [hunk 14]
```diff
@@ -232,17 +223,15 @@ def F(self):
...
-            if self.timezone and self.timezone.dst(self.data):
-                return '1'
-            else:
-                return '0'
-        except Exception:
-            # pytz raises AmbiguousTimeError during the autumn DST change.
-            # This happens mainly when __init__ receives a naive datetime
-            # and sets self.timezone = get_default_timezone().
+        if (not self.timezone or
+                _datetime_ambiguous_or_imaginary(self.data, self.timezone)):
```
Comment (felixxm) [inline django/utils/dateformat.py:227]:
```suggestion
        if (
            not self.timezone or
            _datetime_ambiguous_or_imaginary(self.data, self.timezone)
        ):
```

File: django/utils/dateformat.py [hunk 15]
```diff
@@ -232,17 +223,15 @@ def F(self):
...
             return ''
 
+        if self.timezone.dst(self.data):
+            return '1'
+        else:
+            return '0'
```
Comment (felixxm) [inline django/utils/dateformat.py:233]:
```suggestion
        return '1' if self.timezone.dst(self.data) else '0'
```

File: django/utils/timezone.py (hunk 2 above)
Comment (carltongibson) [reply django/utils/timezone.py]: Pulled this into #13916

File: docs/releases/3.2.txt [hunk 16]
```diff
@@ -657,6 +657,9 @@ MySQL 5.7 and higher.
 Miscellaneous
 -------------
 
+* Support was added for using non-``pytz`` time zones, such as Python 3.9+'s
```
Comment (felixxm) [inline docs/releases/3.2.txt:660]:
```suggestion
* Django now supports non-``pytz`` time zones ....
```

File: docs/topics/i18n/timezones.txt [hunk 17]
```diff
@@ -26,8 +26,15 @@ to this problem is to use UTC in the code and use local time only when
...
 Time zone support is disabled by default. To enable it, set :setting:`USE_TZ =
-True <USE_TZ>` in your settings file. Time zone support uses pytz_, which is
-installed when you install Django.
+True <USE_TZ>` in your settings file. By default, time zone support uses pytz_,
+which is installed when you install Django; Django also supports the use of
+other time zone implementations like :mod:`zoneinfo` by passing
+:class:`~datetime.tzinfo` objects directly to functions in
+:mod:`django.utils.timezone`.
+
+.. versionchanged: 3.2
```
Comment (felixxm) [inline docs/topics/i18n/timezones.txt:35]:
```suggestion
.. versionchanged:: 3.2
```

File: tests/db_functions/datetime/test_extract_trunc.py [hunk 18]
```diff
@@ -1206,94 +1225,98 @@ def test_trunc_func_with_timezone(self):
...
-                ],
-                lambda m: (m.start_datetime, m.truncated),
-            )
+        for melb in self.get_timezones('Australia/Melbourne'):
+            with self.subTest(repr(melb)):
+                def test_datetime_kind(kind):
+                    self.assertQuerysetEqual(
+                        DTModel.objects.annotate(
+                            truncated=Trunc('start_datetime',
+                                            kind, output_field=DateTimeField(), tzinfo=melb)
```
Comment (felixxm) [inline tests/db_functions/datetime/test_extract_trunc.py:1234]: Please use hanging indentation (everywhere).

File: tests/utils_tests/test_timezone.py [hunk 19]
```diff
@@ -1,14 +1,38 @@
...
+    try:
+        from backports import zoneinfo
+    except ImportError:
+        zoneinfo = None
+
 from django.test import SimpleTestCase, override_settings
 from django.utils import timezone
 
-CET = pytz.timezone("Europe/Paris")
+PARIS = pytz.timezone('Europe/Paris')
```
Comment (felixxm) [inline tests/utils_tests/test_timezone.py:18]: Diff will be smaller without this unnecessary change.

File: django/forms/utils.py (hunk 10 above)
Comment (carltongibson) [reply django/forms/utils.py:170]: Yes, this is covered in `tests/timezones/tests.py` `LegacyFormsTests.test_form_with_ambiguous_time()`

File: django/forms/utils.py (hunk 11 above)
Comment (carltongibson) [reply django/forms/utils.py:168]: I swapped the lines: you start with an ambiguous naive datetime. Adding the `tz` and `fold` makes it not so. (So `value`)

Comment (carltongibson) [reply django/forms/utils.py:168]: Not 100% sure. I will follow-up with @pganssle about this. 

Testing you can't get ambiguous results regardless of using `value` or `return_value`, and adjustments of fold: 

```
>>> import datetime
>>> from django.utils import timezone
>>> import zoneinfo
>>> d = datetime.datetime(2015, 3, 29, 2, 30)
>>> d.fold
0
>>> PARIS_ZI = zoneinfo.ZoneInfo('Europe/Paris')
>>> timezone._datetime_ambiguous_or_imaginary(d, PARIS_ZI)
True
>>> a = d.replace(tzinfo=PARIS_ZI)
>>> timezone._datetime_ambiguous_or_imaginary(a, PARIS_ZI)
True
>>> a2 = a.replace(fold=1)
>>> timezone._datetime_ambiguous_or_imaginary(a2, PARIS_ZI)
True
```

**Comment by aaugustin** [comment]
🎉

File: django/forms/utils.py (hunk 11 above)
Comment (pganssle) [reply django/forms/utils.py:166]: Yeah, this is fine. I think the order-swap was to avoid unnecessary work in the common case where the user is using `pytz`, and because frequently people prefer to avoid multiple returns.

If you're OK with multiple returns (I don't think it's really any worse in this case), I think that `_is_pytz_zone` is fairly lightweight — I would guess it's mostly a bit of function call overhead, so this swapped approach seems fine.

File: tests/utils_tests/test_timezone.py (hunk 19 above)
Comment (pganssle) [reply tests/utils_tests/test_timezone.py]: I suppose it's somewhat unnecessary, but it is confusing because `CET` represents standard time, but `pytz.timezone("Europe/Paris")` is both `CEST` and `CET`.  The nomenclature is further confused because `CEST` is used elsewhere to represent the _fixed_ offset of Central European Summer Time.

True, changing the name of this variable has no _instrumental_ value, but I think it improves the readability of the tests, especially considering the other changes accompanying it.

Might be worth a later PR to go through and make the `s/CET/PARIS/g` change in this file again.

**Comment by pganssle** [comment]
Thanks so much for your help with this everyone. This was a great first experience contributing to Django :heart:
